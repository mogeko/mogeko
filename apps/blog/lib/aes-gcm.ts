import { Buffer } from "node:buffer";
import { subtle, webcrypto } from "node:crypto";
import { TextDecoder, TextEncoder } from "node:util";

export async function encrypt(plaintext: string, password: string) {
  const pwUtf8 = new TextEncoder().encode(password);
  const pwHash = await subtle.digest("SHA-256", pwUtf8);

  // Since `webcrypto.getRandomValues` fills the buffer with random bytes,
  // we use `Buffer.allocUnsafe` to allocate RAM directly.
  const iv = webcrypto.getRandomValues(Buffer.allocUnsafe(12));
  const alg = { name: "AES-GCM", iv };
  const key = await subtle.importKey("raw", pwHash, alg, false, ["encrypt"]);

  const ptUint8 = new TextEncoder().encode(plaintext);
  const ctBuffer = await subtle.encrypt(alg, key, ptUint8);

  return iv.toString("hex") + Buffer.from(ctBuffer).toString("base64");
}

export async function decrypt(ciphertext: string, password: string) {
  const pwUtf8 = new TextEncoder().encode(password);
  const pwHash = await subtle.digest("SHA-256", pwUtf8);

  const iv = Buffer.from(ciphertext.slice(0, 24), "hex");
  const alg = { name: "AES-GCM", iv };
  const key = await subtle.importKey("raw", pwHash, alg, false, ["decrypt"]);

  const ctBuffer = Buffer.from(ciphertext.slice(24), "base64");

  const plainBuffer = await subtle.decrypt(alg, key, ctBuffer);

  return new TextDecoder().decode(plainBuffer);
}
