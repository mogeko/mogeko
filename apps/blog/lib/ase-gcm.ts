import { webcrypto } from "node:crypto";

export async function encrypt(plaintext: string, password: string) {
  const pwUtf8 = new TextEncoder().encode(password);
  const pwHash = await webcrypto.subtle.digest("SHA-256", pwUtf8);

  const iv = webcrypto.getRandomValues(new Uint8Array(12));

  const alg = { name: "AES-GCM", iv: iv };

  const key = await webcrypto.subtle.importKey("raw", pwHash, alg, false, [
    "encrypt",
  ]);

  const ptUint8 = new TextEncoder().encode(plaintext);
  const ctBuffer = await webcrypto.subtle.encrypt(alg, key, ptUint8);

  const ctArray = Array.from(new Uint8Array(ctBuffer));
  const ctStr = ctArray.map((byte) => String.fromCharCode(byte)).join("");
  const ctBase64 = Buffer.from(ctStr, "binary").toString("base64");

  const ivHex = Array.from(iv)
    .map((b: number) => `00${b.toString(16)}`.slice(-2))
    .join("");

  return ivHex + ctBase64;
}

export async function decrypt(ciphertext: string, password: string) {
  const pwUtf8 = new TextEncoder().encode(password);
  const pwHash = await webcrypto.subtle.digest("SHA-256", pwUtf8);

  const iv = Buffer.from(ciphertext.slice(0, 24), "hex");

  const alg = { name: "AES-GCM", iv: new Uint8Array(iv) };

  const key = await webcrypto.subtle.importKey("raw", pwHash, alg, false, [
    "decrypt",
  ]);

  const ctStr = Buffer.from(ciphertext.slice(24), "base64").toString("binary");
  const ctUint8 = new Uint8Array(
    (ctStr.match(/[\s\S]/g) as RegExpExecArray).map((ch) => ch.charCodeAt(0)),
  );

  const plainBuffer = await webcrypto.subtle.decrypt(alg, key, ctUint8);
  const plaintext = new TextDecoder().decode(plainBuffer);

  return plaintext;
}
