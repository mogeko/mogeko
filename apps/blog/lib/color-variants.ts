import { type VariantProps, cva } from "class-variance-authority";

export const colorVariants = cva([], {
  variants: {
    color: {
      blue: ["text-blue-700"],
      blue_background: ["bg-blue-300"],
      brown: ["text-amber-700"],
      brown_background: ["bg-amber-300"],
      gray: ["text-gray-700"],
      gray_background: ["bg-gray-300"],
      green: ["text-green-700"],
      green_background: ["bg-green-300"],
      orange: ["text-orange-700"],
      orange_background: ["bg-orange-300"],
      pink: ["text-pink-700"],
      pink_background: ["bg-pink-300"],
      purple: ["text-purple-700"],
      purple_background: ["bg-purple-300"],
      red: ["text-red-700"],
      red_background: ["bg-red-300"],
      yellow: ["text-yellow-700"],
      yellow_background: ["bg-yellow-300"],
      default: null,
      default_background: null,
    },
    defaultVariants: {
      color: "default",
    },
  },
});

export type ColorVariantProps = VariantProps<typeof colorVariants>;
