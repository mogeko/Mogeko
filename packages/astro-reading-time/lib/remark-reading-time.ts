import getReadingTime from "reading-time";
import { toString } from "mdast-util-to-string";
import type { RemarkPlugin } from "@astrojs/markdown-remark";

export const remarkReadingTime: RemarkPlugin<[]> = () => {
  return (tree, { data }) => {
    const readingTime = getReadingTime(toString(tree));

    Object.assign((data.astro as any).frontmatter, {
      minutesRead: readingTime.text,
      wordCount: readingTime.words,
    });
  };
};
