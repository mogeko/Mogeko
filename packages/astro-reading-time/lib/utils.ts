const isObject = (val: any): boolean => val && typeof val === "object";

export function assocPath(path: string[], val: any, obj: Obj): Obj {
  if (path.length === 0) return val;

  const [idx, ...tail] = path;

  return Object.assign(obj, {
    [idx]: assocPath(tail, val, isObject(obj[idx]) ? obj[idx] : {}),
  });
}

type Obj = Record<string, any>;
