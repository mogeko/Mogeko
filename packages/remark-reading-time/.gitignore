# build output
dist/

# dependencies
node_modules/

# testing
coverage/

# turbo
.turbo/

# logs
npm-debug.log*
yarn-debug.log*
yarn-error.log*
pnpm-debug.log*

# environment variables
.env
.env.local
.env.development.local
.env.test.local
.env.production.local
.env.production

# misc
.DS_Store
*.pem
