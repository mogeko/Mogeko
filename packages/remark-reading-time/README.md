# remark-reading-time

A remark plugin to calculate reading time of markdown file.
