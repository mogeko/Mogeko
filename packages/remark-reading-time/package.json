{
  "name": "remark-reading-time",
  "version": "0.0.1",
  "private": true,
  "description": "A remark plugin to calculate reading time of markdown file",
  "type": "module",
  "exports": {
    ".": {
      "import": "./dist/index.js"
    }
  },
  "types": "dist/index.d.ts",
  "keywords": [
    "remark",
    "plugin",
    "reading-time"
  ],
  "author": "Mogeko <zhengjunyi@live.com>",
  "license": "MIT",
  "scripts": {
    "build": "tsup",
    "lint": "prettier --check .",
    "fmt": "prettier --write ."
  },
  "dependencies": {
    "mdast-util-to-string": "^4.0.0",
    "reading-time": "^1.5.0"
  },
  "devDependencies": {
    "@astrojs/markdown-remark": "^4.2.1",
    "prettier": "^3.2.5",
    "tsconfig": "workspace:^",
    "tsup": "^8.0.2",
    "typescript": "^5.3.3"
  }
}
