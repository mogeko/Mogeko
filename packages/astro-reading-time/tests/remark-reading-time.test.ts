import { remarkReadingTime } from "@/lib/remark-reading-time";
import { describe, it, expect } from "vitest";
import remarkParse from "remark-parse";
import remarkStringify from "remark-stringify";
import { unified } from "unified";

const md = `# Hello, world!\n\nThis is a test markdown content.`;

describe("remarkReadingTime", () => {
  it("adds reading time to the frontmatter", async () => {
    const processor = unified()
      .use(remarkParse)
      .use(remarkReadingTime)
      .use(remarkStringify);

    const result = await processor.process(md);

    expect(result.data.astro).toEqual({
      frontmatter: { minutesRead: "1 min read", wordCount: 7 },
    });
    expect(result.data).toMatchSnapshot();
  });
});
