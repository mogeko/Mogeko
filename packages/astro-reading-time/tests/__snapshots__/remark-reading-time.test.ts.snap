// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`remarkReadingTime > adds reading time to the frontmatter 1`] = `
{
  "astro": {
    "frontmatter": {
      "minutesRead": "1 min read",
      "wordCount": 7,
    },
  },
}
`;
